### PR TITLE
[None][fix] mxfp4 padding bug for TRT-LLM and CUTLASS MoE backends

### PR DESCRIPTION
**Pull request by nekorobov**
## Summary by CodeRabbit

* **New Features**
  * Added configurable weight alignment and bias data-type when creating quantized weights.
  * Centralized alignment computation and standardized pre-shard padding for weights and scales across fused MoE backends.

* **Bug Fixes**
  * Eliminated fractional scaling and post-shard padding errors that caused inconsistent shard distributions.
  * Fixed 1D bias and mixed-precision loading behaviors for MXFP4/TRT paths, improving reliability.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by dongfengy** [comment]
My local test results:

```
dp4_CUTLASS.log:[08/25/2025-21:37:32] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 89.39
dp4_CUTLASS.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[dp4-CUTLASS]
dp4_CUTLASS.log:=================== 1 passed, 1 warning in 145.35s (0:02:25) ===================
dp4_TRTLLM.log:[08/25/2025-21:40:05] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 89.69
dp4_TRTLLM.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[dp4-TRTLLM]
dp4_TRTLLM.log:=================== 1 passed, 1 warning in 129.97s (0:02:09) ===================
ep4_CUTLASS.log:[08/25/2025-21:32:11] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 89.92
ep4_CUTLASS.log:0.09s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[ep4-CUTLASS]
ep4_CUTLASS.log:=================== 1 passed, 1 warning in 142.01s (0:02:22) ===================
ep4_TRTLLM.log:[08/25/2025-21:34:43] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.75
ep4_TRTLLM.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[ep4-TRTLLM]
ep4_TRTLLM.log:=================== 1 passed, 1 warning in 129.56s (0:02:09) ===================
tp1_CUTLASS.log:[08/25/2025-21:05:21] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.22
tp1_CUTLASS.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp1-CUTLASS]
tp1_CUTLASS.log:=================== 1 passed, 1 warning in 200.61s (0:03:20) ===================
tp1_TRTLLM.log:[08/25/2025-21:08:49] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.37
tp1_TRTLLM.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp1-TRTLLM]
tp1_TRTLLM.log:=================== 1 passed, 1 warning in 184.94s (0:03:04) ===================
tp2_CUTLASS.log:[08/25/2025-21:11:43] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.14
tp2_CUTLASS.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp2-CUTLASS]
tp2_CUTLASS.log:=================== 1 passed, 1 warning in 151.13s (0:02:31) ===================
tp2_TRTLLM.log:[08/25/2025-21:15:19] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.83
tp2_TRTLLM.log:0.11s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp2-TRTLLM]
tp2_TRTLLM.log:=================== 1 passed, 1 warning in 165.58s (0:02:45) ===================
tp4_CUTLASS.log:[08/25/2025-21:18:56] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.90
tp4_CUTLASS.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp4-CUTLASS]
tp4_CUTLASS.log:=================== 1 passed, 1 warning in 169.14s (0:02:49) ===================
tp4_TRTLLM.log:[08/25/2025-21:22:45] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.14
tp4_TRTLLM.log:0.12s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp4-TRTLLM]
tp4_TRTLLM.log:=================== 1 passed, 1 warning in 179.67s (0:02:59) ===================
tp8_CUTLASS.log:[08/25/2025-21:26:09] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 90.07
tp8_CUTLASS.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp8-CUTLASS]
tp8_CUTLASS.log:=================== 1 passed, 1 warning in 164.34s (0:02:44) ===================
tp8_TRTLLM.log:[08/25/2025-21:29:21] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 89.84
tp8_TRTLLM.log:0.10s setup    accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_w4_4gpus[tp8-TRTLLM]
tp8_TRTLLM.log:=================== 1 passed, 1 warning in 164.16s (0:02:44) ===================
```

**Review by dongfengy — APPROVED** [review]
Code LGTM. It might be possible to eliminate some duplicated code but let's don't worry about it for now.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 1]
```diff
@@ -2085,6 +2085,29 @@ def load_quant_scales(self, module: torch.nn.Module, weights: Dict):
...
+    alignment = lcm(weight_alignment, scaling_vector_size)
+    alignment = lcm(alignment, tp_size)
+
+    # If after the alignment, the sharding dim per shard is not a multiple of weight_alignment,
+    # we need to pad the weights to make it a multiple of weight_alignment.
+    padded_weights_dim = math.ceil(shard_dim_size / alignment) * alignment
+    per_shard = padded_weights_dim // tp_size
+    if per_shard % weight_alignment != 0:
+        alignment = weight_alignment * math.ceil(
+            per_shard / weight_alignment) * tp_size
```
Comment (hlu1) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py:2106]:
```suggestion
    def _align(dim, alignment):
        return math.ceil(dim / alignment) * alignment
    
    return _align(shard_dim_size, tp_size * weight_alignment)
```
This code can be simplified.

**Review by hlu1 — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py (hunk 1 above)
Comment (dongfengy) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py:2106]: Looks like entire function is the same as` math.ceil(shard_dim_size/ (weight_alignment*tp_size)) * weight_alignment*tp_size` if we know shard_dim_size is 2880, tp size is 1 2 4 8, scaling_vector_size is 32, weight_alignment is 128 or 256 because lcms are no op here.
